src/Storages/StorageS3.cpp

```diff
@@ -45,100 +45,217 @@ namespace ErrorCodes
     extern const int UNEXPECTED_EXPRESSION;
     extern const int S3_ERROR;
 }
+class StorageS3Source::DisclosedGlobIterator::Impl
+{
 
+public:
+    Impl(Aws::S3::S3Client & client_, const S3::URI & globbed_uri_)
+        : client(client_), globbed_uri(globbed_uri_) {
 
-namespace
-{
-    class StorageS3Source : public SourceWithProgress
+        std::lock_guard lock(mutex);
+
+        if (globbed_uri.bucket.find_first_of("*?{") != globbed_uri.bucket.npos)
+            throw Exception("Expression can not have wildcards inside bucket name", ErrorCodes::UNEXPECTED_EXPRESSION);
+
+        const String key_prefix = globbed_uri.key.substr(0, globbed_uri.key.find_first_of("*?{"));
+
+        if (key_prefix.size() == globbed_uri.key.size())
+            buffer.emplace_back(globbed_uri.key);
```

Comment:
Shouldn't we set `is_finished` here?

Reply:
And also init `buffer_iter`

Reply:
I suppose we need to init `buffer_iter` with `buffer.begin()` anyway

Reply:
Ah, we call fillInternalBufferAssumeLocked. Ok.
But then why do we need this if?

Reply:
And why do we call `fillInternalBufferAssumeLocked` here ...

Reply:
We have fast path if case of absent asterisks (we don't have to list anything, just return the current bucket + file). Fixed.